### PR TITLE
Add foreman_domain

### PR DESCRIPTION
**Pull request by m-bucher**

### CODE REVIEW

File: modules/foreman_domain.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+    try:
+        create_server(server_url, (username, password), verify_ssl)
+    except Exception as e:
+        module.fail_json(msg="Failed to connect to Foreman server: %s " % e)
+
+    ping_server(module)
+    try:
+        # Try to find the Domain to work on
+        entity = None
```
Comment (sean797) [inline modules/foreman_domain.py]: I don't think you need this.

File: modules/foreman_domain.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            server_url=dict(required=True),
+            username=dict(required=True),
+            password=dict(required=True, no_log=True),
+            verify_ssl=dict(type='bool', default=True),
+            name=dict(required=True),
+            description=dict(),
+            dns_proxy=dict(),
+            locations=dict(type='list'),
+            organizations=dict(type='list'),
+            state=dict(required=True, choices=['present', 'absent']),
```
Comment (sean797) [inline modules/foreman_domain.py]: default to `present` please.

File: test/test_playbooks/domain.yml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+- hosts: tests
+  vars:
+    auth_params: &auth_params
+      username: "{{ foreman_username }}"
+      password: "{{ foreman_password }}"
+      server_url: "{{ foreman_server_url }}"
+      verify_ssl: "{{ foreman_verify_ssl }}"
+    test_data: &test_data
+      name: "example.org"
+      #dns_proxy: "foreman.example.com"
```
Comment (sean797) [inline test/test_playbooks/domain.yml]: ?? It would be nice if a test included setting `dns_proxy`

File: test/test_playbooks/domain.yml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      msg: "Create domain did not report expected change!"
+    when: result.changed != true
+  - name: "Create domain again"
+    foreman_domain:
+      <<: *auth_params
+      <<: *test_data
+      state: "present"
+    register: result
+  - fail:
+      msg: "Create domain did not report expected change!"
```
Comment (sean797) [inline test/test_playbooks/domain.yml]: s/did not report expected/reported/

File: test/test_playbooks/domain.yml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+      msg: "Delete domain did not report expected change!"
+    when: result.changed != true
+  - name: "Delete domain again"
+    foreman_domain:
+      <<: *auth_params
+      <<: *test_data
+      state: "absent"
+    register: result
+  - fail:
+      msg: "Delete domain did not report expected change!"
```
Comment (sean797) [inline test/test_playbooks/domain.yml]: s/did not report expected/reported/

File: module_utils/ansible_nailgun_cement.py [hunk 6]
```diff
@@ -293,6 +300,11 @@ def find_repository_set(module, name, product, failsafe=False):
     return handle_find_response(module, repo_set.search(), message="No repository set found for %s" % name, failsafe=failsafe)
 
 
+def find_smart_proxy(module, name, failsafe=False):
+    smart_proxy = SmartProxy().search(query={'search': 'name="{}"'.format(name)})
+    return handle_find_response(module, smart_proxy, message="No SmartProxy found for %s" % name, failsafe=failsafe)
```
Comment (akofink) [inline module_utils/ansible_nailgun_cement.py]: s/SmartProxy/Smart Proxy/

File: test/test_playbooks/domain_vars.yml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+---
+domain_name: "example.org"
+domain_dns_proxy: "centos7-devel.example.com"
+domain_locations:
+  - 'Default Location'
+domain_organizations:
+  - 'Default Organization'
```
Comment (mdellweg) [inline test/test_playbooks/domain_vars.yml]: Please use the provided "Test <Taxonomy>" here.

File: test/test_playbooks/domain.yml [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  - include_tasks: tasks/location.yml
+    vars:
+      location_name: "{{ item }}"
+      location_state: "present"
+    with_items: "{{ domain_locations }}"
+  - include_tasks: tasks/organization.yml
+    vars:
+      organization_name: "{{ item }}"
+      organization_state: "present"
+    with_items: "{{ domain_organizations }}"
```
Comment (mdellweg) [inline test/test_playbooks/domain.yml]: Can you clean up the entities again afterwards?

File: test/test_playbooks/domain_vars.yml [hunk 9]
```diff
@@ -0,0 +1,8 @@
+---
```
Comment (mdellweg) [inline test/test_playbooks/domain_vars.yml]: Is there a net benefit of putting this in an extra file?

File: modules/foreman_domain.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+    try:
+        # Try to find the Domain to work on
+        entity = find_domain(module, name=domain_dict['name'], failsafe=True)
+    except Exception as e:
+        module.fail_json(msg='Failed to find entity: %s ' % e)
+
+    if 'dns_proxy' in domain_dict:
+        domain_dict['dns_proxy'] = find_smart_proxy(module, domain_dict['dns_proxy'])
+
+    if 'locations' in domain_dict:
```
Comment (akofink) [inline modules/foreman_domain.py]: You could implement `find_locations` in ansible_nailgun_cement like [this](https://github.com/theforeman/foreman-ansible-modules/blob/master/module_utils/ansible_nailgun_cement.py#L282), or use [`find_entities`](https://github.com/theforeman/foreman-ansible-modules/blob/master/module_utils/ansible_nailgun_cement.py#L157). Same for organizations.

File: test/test_playbooks/domain_vars.yml (hunk 9 above)
Comment (m-bucher) [reply test/test_playbooks/domain_vars.yml]: I use them for both hosts, fixtures and tests.
I am not sure if I can add variables that both hosts can use within the playbook.

File: test/test_playbooks/domain.yml (hunk 8 above)
Comment (m-bucher) [reply test/test_playbooks/domain.yml]: I will have to think/learn about how to save the _change_-state in a variable to make sure that only entities that have not existed before are deleted

File: test/test_playbooks/domain_vars.yml (hunk 9 above)
Comment (mdellweg) [reply test/test_playbooks/domain_vars.yml]: Nice catch.

File: test/test_playbooks/domain.yml (hunk 8 above)
Comment (mdellweg) [reply test/test_playbooks/domain.yml]: That would even raise the standards for our test_playbooks. ;)

File: test/test_playbooks/domain.yml [hunk 11]
```diff
@@ -0,0 +1,47 @@
+---
+- hosts: fixtures
+  gather_facts: false
+  tasks:
+  - name: Load server config
+    include_vars:
+      file: server_vars.yml
+  - name: Load domain test-config
+    include_vars:
+      file: domain_vars.yml
```
Comment (akofink) [inline test/test_playbooks/domain.yml]: +1 to this design

File: modules/foreman_domain.py (hunk 10 above)
Comment (m-bucher) [reply modules/foreman_domain.py]: Agreed, I added `find_locations` and `find_organizations` similar to `find_repositories`

File: test/test_playbooks/domain_vars.yml (hunk 7 above)
Comment (m-bucher) [reply test/test_playbooks/domain_vars.yml]: I'd rather keep all config-data for the test in one file.

File: test/test_playbooks/domain.yml (hunk 8 above)
Comment (m-bucher) [reply test/test_playbooks/domain.yml]: Went the easy way by just doing a delete (for now :-) )

**Comment by mdellweg** [comment]
@sean797 i think, your concerns were addressed...

**Pull request merged**